src/System.Text.Json/tests/JsonObjectTests.cs

```diff
@@ -0,0 +1,609 @@
+            Assert.IsType<JsonString>(phoneNumbers["work"]);
+            Assert.IsType<JsonString>(phoneNumbers["home"]);
+
+            Assert.IsType<JsonObject>(person["addresses"]);
+            var addresses = person["addresses"] as JsonObject;
+            Assert.IsType<JsonObject>(addresses["office"]);
+            Assert.IsType<JsonObject>(addresses["home"]);
+        }
+
+        [Fact]
+        public static void TestAssignmentDefinition()
+        {
+            JsonNode employee = EmployeesDatabase.GetNextEmployee().Value;
+            Assert.IsType<JsonObject>(employee);
+        }
+
+        [Fact]
+        public static void TestAddingKeyValuePair()
+        {
+            var employees = new JsonObject
+            {
+                EmployeesDatabase.GetNextEmployee(),
+                EmployeesDatabase.GetNextEmployee(),
+                EmployeesDatabase.GetNextEmployee(),
+                EmployeesDatabase.GetNextEmployee(),
+            };
+
+            string prevId = "";
+            foreach ((string id, JsonNode employee) in employees)
+            {
+                Assert.NotEqual(prevId, id);
+                prevId = id;
+
+                Assert.IsType<JsonObject>(employee);
+            }
+        }
+
+        [Fact]
+        public static void TestAddingKeyValuePairAfterInitialization()
+        {
+            var employees = new JsonObject();
+            foreach (KeyValuePair<string, JsonNode> employee in EmployeesDatabase.GetTenBestEmployees())
+            {
+                employees.Add(employee);
+            }
+
+            string prevId = "";
+            foreach ((string id, JsonNode employee) in employees)
+            {
+                Assert.NotEqual(prevId, id);
+                prevId = id;
+
+                Assert.IsType<JsonObject>(employee);
+            }
+        }
+
+        [Fact]
+        public static void TestAddingKeyValuePairsCollection()
+        {
+            var employees = new JsonObject(EmployeesDatabase.GetTenBestEmployees());
+
+            string prevId = "";
+            foreach ((string id, JsonNode employee) in employees)
+            {
+                Assert.NotEqual(prevId, id);
+                prevId = id;
+
+                Assert.IsType<JsonObject>(employee);
+            }
+        }
+
+        [Fact]
+        public static void TestAddingKeyValuePairsCollectionAfterInitialization()
+        {
+            var employees = new JsonObject();
+            employees.AddRange(EmployeesDatabase.GetTenBestEmployees());
+
+            string prevId = "";
+            foreach ((string id, JsonNode employee) in employees)
+            {
+                Assert.NotEqual(prevId, id);
+                prevId = id;
+
+                Assert.IsType<JsonObject>(employee);
+            }
+        }
+
+        [Fact]
+        public static void TestContains()
+        {
+            var person = new JsonObject
+            {
+                { "name", "John" },
```

Comment:
What happens when testing Contains when there are multiple properties of the same name?

What about if they are escaped? Add those tests.

Reply:
Such `JsonObjects` are impossible to create.

Reply:
> What about if they are escaped? 

What do you mean exactly?

Reply:
> What do you mean exactly?

```C#
{
	"name": "John",
	"n\\u0061me": "John"
}
```

Are "name" and "n\\u0061me" considered identical? \\u0061 is 'a' but written in expanded form.

Reply:
Ok, I see, thanks,  I will test it.

Reply:
I'm assuming that the API operates on C# strings, not JSON strings, so the Add method and/or indexer would treat the two as different strings.

If they were parsed out of the payload (and it was only one backslash) then the overwrite/throw/ignore behavior would kick in.

Reply:
> Such `JsonObjects` are impossible to create.

Why is that the case? Is it because we are using Dictionary under the covers? Having duplicate property names is valid JSON (though generally discouraged due to interoperability):
https://tools.ietf.org/html/rfc8259#section-4
>  An object whose names are all unique is interoperable in the sense
   that all software implementations receiving that object will agree on
   the name-value mappings.  When the names within an object are not
   unique, the behavior of software that receives such an object is
   unpredictable.  Many implementations report the last name/value pair
   only.  Other implementations report an error or fail to parse the
   object, and some implementations report all of the name/value pairs,
   including duplicates.

Also, the string comparison section might be a useful read:
https://tools.ietf.org/html/rfc8259#section-8.3